package/MDAnalysis/analysis/base.py

```diff
@@ -652,3 +654,236 @@ def _filter_baseanalysis_kwargs(function, kwargs):
+
+        Returns
+        -------
+        computation_groups : list[np.ndarray]
+            list of (n, 2) shaped np.ndarrays with frame indices and numbers
+        """
+        if frames is None:
+            start, stop, step = self._trajectory.check_slice_indices(start, stop, step)
+            used_frames = np.arange(start, stop, step)
+        elif not all(opt is None for opt in [start, stop, step]):
+            raise ValueError("start/stop/step cannot be combined with frames")
+        else:
+            used_frames = frames
+
+        if all((isinstance(obj, bool) for obj in used_frames)):
+            arange = np.arange(len(used_frames))
+            used_frames = arange[used_frames]
+
+        # this numpy thing is similar to list(enumerate(frames))
+        enumerated_frames = np.vstack([np.arange(len(used_frames)), used_frames]).T
+
+        return np.array_split(enumerated_frames, n_parts)
+
+
+    def _compute(self, indexed_frames: np.ndarray, verbose=None, *, progressbar_kwargs={}) -> "AnalysisBase":
+        """Perform the calculation on a balanced slice of frames
+        that have been setup prior to that using _setup_computation_groups()
+
+        Parameters
+        ----------
+        indexed_frames : np.ndarray
+            np.ndarray of (n, 2) shape,
+            where first column is frame iteration indices
+            and second is frame numbers
+
+        verbose : bool, optional
+            Turn on verbosity
+
+        progressbar_kwargs : dict, optional
+            ProgressBar keywords with custom parameters regarding progress bar position, etc;
+            see :class:`MDAnalysis.lib.log.ProgressBar` for full list.
+
+
+        .. versionchanged:: 2.2.0
+            Added ability to analyze arbitrary frames by passing a list of
+            frame indices in the `frames` keyword argument.
+
+        .. versionchanged:: 2.5.0
+            Add `progressbar_kwargs` parameter,
+            allowing to modify description, position etc of tqdm progressbars
+        """
+        logger.info("Choosing frames to analyze")
+        # if verbose unchanged, use class default
+        verbose = getattr(self, "_verbose", False) if verbose is None else verbose
+
+        logger.info("Starting preparation")
+        logger.info("Starting analysis loop over %d trajectory frames", self.n_frames)
+
+        if len(indexed_frames) == 0:  # if `frames` were empty in `run` or `stop=0`
+            return self
+
+        frame_indices, frames = (
+            indexed_frames[:, 0],
+            indexed_frames[:, 1],
+        )
+        trajectory = self._trajectory[frames]
+        for idx, ts in enumerate(ProgressBar(trajectory, verbose=verbose, **progressbar_kwargs)):
+            i = frame_indices[idx]
+            self._frame_index = i
+            self._ts = ts
+            self.frames[i] = ts.frame
+            self.times[i] = ts.time
+            self._single_frame()
+        logger.info("Finishing up")
+        return self
```

Comment:
Could we refactor the inner loop so that the same code is shared between `AnalysisBase` and `ParallelAnalysisBase`?

I am trying to get a sense if we can mitigate the class proliferation problem by keeping as much code as possible shared.